lib/ansible/plugins/callback/mqtt.py

```diff
@@ -0,0 +1,333 @@
+# Copyright IBM Corp. 2017
+# Author(s): Matthew Treinish <mtreinish@kortar.org>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import datetime
+import json
+import os
+import socket
+import uuid
+
+import yaml
+
+from ansible.plugins.callback import CallbackBase
+
+HAS_PAHO = True
+try:
+    from paho.mqtt import publish
+except ImportError:
+    HAS_PAHO = False
+
+
+class CallbackModule(CallbackBase):
+    """
+    This ansible callback plugin is for sending status updates to mqtt
+    during playbook execution.
+
+    This plugin makes use of the following env variables:
+        MQTT_HOSTNAME   (required): The hostname of the MQTT broker
+        MQTT_BASE_TOPIC (optional): The base topic to use for messages,
+                                    defaults to 'ansible'
+        MQTT_PORT       (optional): The MQTT broker port number, defaults to
+                                    1883
+        MQTT_USERNAME   (optional): Username to authenticate against the broker
+        MQTT_PASSWORD   (optional): Password for MQTT_USERNAME to authenticate
+                                    against the broker
+        MQTT_CA_CERTS   (optional): The path of the Certificate Authority cert
+                                    files to be trusted by the client.
+        MQTT_CERTFILE   (optional): The path pointing to the PEM encoded client
+                                    certificate.
+        MQTT_KEYFILE    (optional): The path pointing to the PEM encoded client
+                                    private key
+        MQTT_CLIENT_ID  (optional): MQTT client identifier, defaults to
+                                    hostname + pid
```

Comment:
This callback needs to be updated to utilize "config".  See other current callback plugins for their `DOCUMENTATION` and `set_options` method.